std-bits/table/src/main/java/org/enso/table/data/column/operation/StorageIterators.java

```diff
@@ -487,4 +582,238 @@ public static <T> ColumnStorage<T> mapOverBooleanStorage(
     }
     return builder.seal();
   }
+
+  @FunctionalInterface
+  public interface ZipOperation<R, S, T> {
+    T apply(long index, R value1, S value2);
+  }
+
+  @FunctionalInterface
+  public interface LongZipOperation<T> {
+    T apply(long index, long value1, boolean isNothing1, long value2, boolean isNothing2);
+  }
+
+  @FunctionalInterface
+  public interface DoubleZipOperation<T> {
+    T apply(long index, double value1, boolean isNothing1, double value2, boolean isNothing2);
+  }
+
+  /**
+   * Zips two storages together, applying an operation to each pair of values. The operation's
+   * result is appended to the builder. The builderConstructor will be passed the expected size to
+   * create a new builder. If skipNothing is true, then if either value is Nothing, the result will
+   * be Nothing and appended automatically.
+   *
+   * @param source1 the first source storage to read from and iterate over.
+   * @param source2 the second source storage to read from and iterate over.
+   * @param builderConstructor a function to create a new builder of the correct type.
+   * @param skipNothing if true, then if either value is Nothing, the result will be Nothing.
+   * @param operation a callback to process a pair of values.
+   * @return a built ColumnStorage from sealing the builder.
+   * @param <R> Input Java type for the first source.
+   * @param <S> Input Java type for the second source.
+   * @param <T> Output Java type for the storage.
+   */
+  public static <R, S, T> ColumnStorage<T> zipOverStorages(
+      ColumnStorage<R> source1,
+      ColumnStorage<S> source2,
+      LongFunction<BuilderForType<T>> builderConstructor,
```

Comment:
nitpick, feel free to ignore

but seems like it may be easier to read if we have
```
@FunctionalInterface
public interface BuilderFactory<T> {
    BuilderForType<T> build(long initialCapacity);
}
```

and then
```suggestion
      BuilderFactory<T> builderFactory,
```